### PR TITLE
Add power-off restart logic for IndexService and IndexNode

### PR DESCRIPTION
**Pull request by xiaocai2333**
Add power-off restart logic for IndexService and IndexNode

Signed-off-by: <xiaocai2333> <cai.zhang@zilliz.com>

### CODE REVIEW

File: internal/indexnode/indexnode.go [hunk 1]
```diff
@@ -197,19 +220,20 @@ func (i *IndexNode) DropIndex(ctx context.Context, request *indexpb.DropIndexReq
...
 	log.Debug("IndexNode", zap.Any("The index of the IndexBuildIDs to be deleted", indexBuildIDs))
 	for _, indexBuildID := range indexBuildIDs {
-		nty := &indexpb.NotifyBuildIndexRequest{
-			Status: &commonpb.Status{
-				ErrorCode: commonpb.ErrorCode_Success,
-			},
-			IndexBuildID:   indexBuildID,
-			NodeID:         Params.NodeID,
-			IndexFilePaths: []string{},
+		metaPath := "/indexes/" + strconv.FormatInt(indexBuildID, 10)
```
Comment (czs007) [inline internal/indexnode/indexnode.go:223]: etcd kv root path should be xxx/indexes/

File: internal/indexnode/task.go [hunk 2]
```diff
@@ -131,30 +137,39 @@ func (it *IndexBuildTask) PostExecute(ctx context.Context) error {
...
+	indexMeta.State = commonpb.IndexState_Finished
 	if it.err != nil {
-		nty.Status.ErrorCode = commonpb.ErrorCode_BuildIndexError
+		indexMeta.State = commonpb.IndexState_Failed
 	}
-
-	ctx = context.TODO()
-	resp, err := it.serviceClient.NotifyBuildIndex(ctx, nty)
+	log.Debug("IndexNode", zap.Any("MetaPath", it.req.MetaPath))
+	err = it.etcdKV.CompareVersionAndSwap(it.req.MetaPath, resp.Kvs[0].Version,
```
Comment (czs007) [inline internal/indexnode/task.go:162]: Fail to update the version does not mean failure. Please consider each situation carefully.

File: internal/indexnode/indexnode.go (hunk 1 above)
Comment (xiaocai2333) [reply internal/indexnode/indexnode.go]: There is no need to slove it temporarily, I will fix it soon.

File: internal/indexnode/task.go (hunk 2 above)
Comment (xiaocai2333) [reply internal/indexnode/task.go]: Yes, I have fully considered what will go wrong.

File: internal/indexnode/task.go [hunk 3]
```diff
@@ -113,6 +119,31 @@ func (it *IndexBuildTask) OnEnqueue() error {
...
+		return err
+	}
+	if indexMeta.Version > it.req.Version || indexMeta.State == commonpb.IndexState_Finished {
+		log.Debug("IndexNode", zap.Any("This task no longer needs to be done with IndexBuildID", indexMeta.IndexBuildID))
+		return errors.New("This task no longer needs to be done with IndexBuildID ")
+	}
+	if indexMeta.MarkDeleted {
+		indexMeta.State = commonpb.IndexState_Finished
+		value := proto.MarshalTextString(&indexMeta)
+		err := it.etcdKV.Save(it.req.MetaPath, value)
```
Comment (czs007) [inline internal/indexnode/task.go:142]: What if save failed?

File: internal/indexnode/task.go [hunk 4]
```diff
@@ -131,29 +162,35 @@ func (it *IndexBuildTask) PostExecute(ctx context.Context) error {
...
-	}
-	if it.err != nil {
-		nty.Status.ErrorCode = commonpb.ErrorCode_BuildIndexError
-	}
-
-	ctx = context.TODO()
-	resp, err := it.serviceClient.NotifyBuildIndex(ctx, nty)
-	if err != nil {
-		log.Warn("indexnode", zap.String("error", err.Error()))
+	fn := func() error {
```
Comment (czs007) [inline internal/indexnode/task.go:165]: The logic of this part is the same as that in PreExecute. Can the code be reused?

File: internal/indexnode/task.go [hunk 5]
```diff
@@ -322,7 +359,26 @@ func (it *IndexBuildTask) Execute(ctx context.Context) error {
 
 			savePath := getSavePathByKey(key)
 
-			err := saveBlob(savePath, value)
+			saveIndexFileFn := func() error {
+				resp, err := it.etcdKV.LoadWithPrefix2(it.req.MetaPath)
```
Comment (czs007) [inline internal/indexnode/task.go:363]: Can it be reused?

File: internal/indexnode/task.go [hunk 6]
```diff
@@ -131,29 +162,35 @@ func (it *IndexBuildTask) PostExecute(ctx context.Context) error {
...
+			log.Debug("IndexService", zap.Any("load meta error with path", it.req.MetaPath))
+			log.Debug("IndexService", zap.Any("Load meta error", err))
+			return err
+		}
+		err = proto.UnmarshalText(string(resp.Kvs[0].Value), &indexMeta)
+		if err != nil {
+			log.Debug("IndexService", zap.Any("Unmarshal error", err))
+			return err
+		}
+		if indexMeta.Version > it.req.Version || indexMeta.State == commonpb.IndexState_Finished {
```
Comment (czs007) [inline internal/indexnode/task.go:178]: should also check MarkDelete

File: internal/indexservice/meta_table.go [hunk 7]
```diff
@@ -15,27 +15,44 @@ import (
...
 	lock sync.RWMutex
 }
 
-func NewMetaTable(kv kv.TxnKV) (*metaTable, error) {
+func NewMetaTable(kv *etcdkv.EtcdKV, i *IndexService) (*metaTable, error) {
 	mt := &metaTable{
-		client: kv,
-		lock:   sync.RWMutex{},
+		client:       kv,
+		indexService: i,
```
Comment (czs007) [inline internal/indexservice/meta_table.go:54]: Don't put indexservice here

File: internal/indexservice/meta_table.go [hunk 8]
```diff
@@ -46,31 +63,74 @@ func NewMetaTable(kv kv.TxnKV) (*metaTable, error) {
...
 
-	_, values, err := mt.client.LoadWithPrefix("indexes")
+	resp, err := mt.client.LoadWithPrefix2(key)
 	if err != nil {
 		return err
 	}
+	var indexBuildIDs []UniqueID
 
-	for _, value := range values {
+	for _, kv := range resp.Kvs {
```
Comment (czs007) [inline internal/indexservice/meta_table.go:76]: Move this loop outside etcd

File: internal/indexnode/task.go (hunk 3 above)
Comment (xiaocai2333) [reply internal/indexnode/task.go]: It should retry.

File: internal/indexnode/task.go (hunk 4 above)
Comment (xiaocai2333) [reply internal/indexnode/task.go:121]: Of course.

File: internal/indexnode/task.go (hunk 5 above)
Comment (xiaocai2333) [reply internal/indexnode/task.go]: Yep.

File: internal/indexservice/meta_table.go (hunk 7 above)
Comment (xiaocai2333) [reply internal/indexservice/meta_table.go]: Alright.

File: internal/indexservice/meta_table.go (hunk 8 above)
Comment (xiaocai2333) [reply internal/indexservice/meta_table.go]: This loop should be kept, and the logic for checking if indexnode is alive should move.

**Review by czs007 — APPROVED** [review]

**Pull request merged**